tracer/src/Datadog.Trace/Debugger/Symbols/SymbolsUploader.cs

```diff
@@ -0,0 +1,288 @@
+                }
+
+                if (!_alreadyProcessed.Add(assembly.Location))
+                {
+                    return;
+                }
+
+                await UploadAssemblySymbols(assembly).ConfigureAwait(false);
+            }
+            catch (Exception e)
+            {
+                Log.Error(e, "Error while trying to extract assembly symbol {Assembly}", assembly);
+            }
+        }
+
+        private async Task UploadAssemblySymbols(Assembly assembly)
+        {
+            using var symbolsExtractor = SymbolExtractor.Create(assembly);
+            if (symbolsExtractor == null)
+            {
+                return;
+            }
+
+            var assemblyScope = symbolsExtractor.GetAssemblySymbol();
+
+            var root = new Root
+            {
+                Service = _serviceName,
+                Env = _environment,
+                Language = "dotnet",
+                Version = _serviceVersion,
+                Scopes = new[] { assemblyScope }
+            };
+
+            await UploadClasses(root, symbolsExtractor.GetClassSymbols()).ConfigureAwait(false);
+        }
+
+        private async Task UploadClasses(Root root, IEnumerable<Model.Scope?> classes)
+        {
+            var count = 0;
+            var builder = StringBuilderCache.Acquire(StringBuilderCache.MaxBuilderSize);
+
+            try
+            {
+                foreach (var classSymbol in classes)
+                {
+                    if (classSymbol == null)
+                    {
+                        continue;
+                    }
+
+                    count += SerializeClass(classSymbol.Value, builder);
+                    if (count < _sizeLimit)
+                    {
+                        continue;
+                    }
+
+                    await Upload(root, builder).ConfigureAwait(false);
+                    count = 0;
+                }
+
+                if (count > 0)
+                {
+                    await Upload(root, builder).ConfigureAwait(false);
+                }
+            }
+            catch (Exception)
+            {
+                if (builder != null)
+                {
+                    StringBuilderCache.Release(builder);
+                }
+            }
+        }
+
+        private async Task Upload(Root root, StringBuilder builder)
+        {
+            FinalizeSymbolForSend(root, builder);
+            await SendSymbol(StringBuilderCache.GetStringAndRelease(builder)).ConfigureAwait(false);
+            ResetPayload();
+        }
+
+        private void ResetPayload()
+        {
+            if (_payload != null)
+            {
+                Array.Clear(_payload, 0, _payload.Length);
+            }
+        }
+
+        private async Task<bool> SendSymbol(string symbol)
+        {
+            var count = Encoding.UTF8.GetByteCount(symbol);
+            if (_payload == null || count >= _payload.Length)
+            {
+                _payload = new byte[count];
```

Comment:
Thread-safety alert: `SymbolsUploder` can be used in multi-threading (the assembly load event can fire at any point in time) and playing with a field that is used by these different crunching threads may cause undesired behavior.
This comment is valid for all the points where `_payload` is in use.

Reply:
The code is async\await but not multithreaded so should not be an issue.

Reply:
https://github.com/DataDog/dd-trace-dotnet/pull/4412#discussion_r1295796329